### PR TITLE
Assignment 3 @VeraKorchemnaya @AndrewKorchemniy

### PR DESCRIPTION
**Pull request by VeraKorchemnaya**
We are doing option 1, building off of existing Assignment-3 branch.

[Link to website](https://brave-stone-07b629c1e.3.azurestaticapps.net/)
[Link to Word endpoint](https://wordlemindbender.azurewebsites.net/Word)
[Link to Top Ten Players endpoint](https://wordlemindbender.azurewebsites.net/Player/GetTopTenPlayers)

### CODE REVIEW

**Review by StrangeRanger — COMMENTED** [review]
## Instructions

**1. Create a leaderboard API Controller** ❌✅
- Create an endpoint that returns the top 10 scores (HttpGet) ✅
  - Results should include Name, average guesses, and number of games played ✅
- Create an endpoint that allows for posting a score (HttpPost) with data in the body ✅
  - A new name should add a new record ✅
  - An existing name should update that record ✅
- Save and load all data with Entity Framework to an Azure SQL database or LocalDb for debugging ✅
  - Class should be named "Player" and "Players" for the DbSet ✅
  - Fields should be: 
    - int PlayerId ✅
    - string Name ✅
    - int GameCount ✅
    - double AverageAttempts ✅
    - int AverageSecondsPerGame ✅(OPTIONAL)
- The logic should reside in a service that is injected via dependency injection into the controller ✅
- Set up CORS (Cross Origin Resource Sharing) ✅

**2. Create a dialog for getting the user's name** ✅
- If we don't have the user's name, present the user with a dialog that allows them to type in their name ✅
- The users current name should show on the game page (upper right) ❌✅
  - ***REVIEW NOTE: Isn't displayed on the game page, but in the header instead.***
  - This should be visible when setting/changing the name. It should change as the user types ✅
- The user should be able to click on their name to change it (with the same dialog) ✅
- Save the name in local storage and load it for the next game if it is set ❌✅
  -  ***REVIEW NOTES: You save it using cookies instead, though I realistically don't think you'll be docked down for this.***
- If the name is not set, display Guest as the name. ✅
  - Clicking on this allows for preemptive setting of the name before the end of the game ✅
- Dialog must be persistent, doesn't close by clicking off the dialog ✅

**3. Save the Score and Seconds to complete** ✅
- Save the score and time to complete in seconds at the end of the game if we have a name. ✅
- If the user hasn't set a name, save the score and seconds as "Guest". (Be sure to prompt first) ✅
- Save should be done with an Axios Post ✅

**4. Create a Leaderboard Page** ✅
- Create a /leaderboard page in the application that shows the top scores ✅
- Use Axios to call the leaderboard API ✅
- Make a nice display of the returned data ✅
- Add a link on the front page to access the leaderboard ✅
- Add a link to the app bar to access the leaderboard ✅

**5. Unit testing** ❌
- Tests for the service ❌
- Integration tests for the controller ❌ (OPTIONAL)

**6. Deploy site to Azure** ✅
- Deploy API site with Github Action to Azure Web App ✅
- Unit Tests should run on build ✅
- Add API Base URL replacement to front end project via secret ❌✅
  - ***REVIEW NOTE: I'm not sure about this one.***


## Extra Credit

- Unit testing on the client side (3) ❌
- Animate the leaderboard (3) ❌
- Create a logo (3) ❌
- Add Average Seconds to win in score (3) ✅

File: .github/workflows/assignment-3_wordlemindbender.yml [hunk 1]
```diff
@@ -1,12 +1,12 @@
 # Docs for the Azure Web Apps Deploy action: https://github.com/Azure/webapps-deploy
 # More GitHub Actions for Azure: https://github.com/Azure/actions
```
Comment (DaisyPalencia) [inline .github/workflows/assignment-3_wordlemindbender.yml:3]: Workflow is correct

**Review by DaisyPalencia — COMMENTED** [review]
Instructions

1. Create a leaderboard API Controller ✅

Create an endpoint that returns the top 10 scores (HttpGet) ✅
Results should include Name, average guesses, and number of games played ✅
Create an endpoint that allows for posting a score (HttpPost) with data in the body ✅
A new name should add a new record ✅
An existing name should update that record ✅
Save and load all data with Entity Framework to an Azure SQL database or LocalDb for debugging ✅
Class should be named "Player" and "Players" for the DbSet ✅
Fields should be:
int PlayerId ✅
string Name ✅
int GameCount ✅
double AverageAttempts ✅
int AverageSecondsPerGame ✅(OPTIONAL)
The logic should reside in a service that is injected via dependency injection into the controller ✅
Set up CORS (Cross Origin Resource Sharing) ✅
2. Create a dialog for getting the user's name ✅

If we don't have the user's name, present the user with a dialog that allows them to type in their name ✅
The users current name should show on the game page (upper right) ✅
This should be visible when setting/changing the name. It should change as the user types ✅
The user should be able to click on their name to change it (with the same dialog) ✅
Save the name in local storage and load it for the next game if it is set ✅
If the name is not set, display Guest as the name. ✅
Clicking on this allows for preemptive setting of the name before the end of the game ✅
Dialog must be persistent, doesn't close by clicking off the dialog ✅
3. Save the Score and Seconds to complete ✅

Save the score and time to complete in seconds at the end of the game if we have a name. ✅
If the user hasn't set a name, save the score and seconds as "Guest". (Be sure to prompt first) ✅
Save should be done with an Axios Post ✅
4. Create a Leaderboard Page ✅

Create a /leaderboard page in the application that shows the top scores ✅
Use Axios to call the leaderboard API ✅
Make a nice display of the returned data ✅
Add a link on the front page to access the leaderboard ✅
Add a link to the app bar to access the leaderboard ✅
5. Unit testing ❌(OPTIONAL) (Not covered in class, but in instructor's repo: https://github.com/granterickson/EWU-CSCD379-2023-Spring)

Tests for the service ❌(OPTIONAL)
Integration tests for the controller ❌(OPTIONAL)
6. Deploy site to Azure ✅

Deploy API site with Github Action to Azure Web App ✅
Unit Tests should run on build ✅
Add API Base URL replacement to front end project via secret ✅
Extra Credit

Unit testing on the client side (3) ❌
Animate the leaderboard (3) ❌
Create a logo (3) ❌
Add Average Seconds to win in score (3) ✅

**Comment by VeraKorchemnaya** [comment]
> 

@StrangeRanger 
Sorry for the confusion, that was leftover from when Meg did the inverse theme! Yes it's very confusing!
Also thank you for pointing out that the user's name isn't displayed on the game page, but instead in the header. I did not even notice that! I think, however, that I'm going to leave it as it makes more sense to me from a UX prospective that the username should be visible on all pages. And about that base URL that just means that I replaced the original API service public profile with a secret that contains this profile (AZUREAPPSERVICE_PUBLISHPROFILE_CB1383D70FE94E1FA90BD298F2110BA4 -> AZUREAPPSERVICE_PUBLISHPROFILE_API)

File: Wordle.Api/Wordle.Api/Controllers/PlayerController.cs [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    {
+        private readonly PlayerService _playerService;
+
+        public PlayerController(PlayerService playerService)
+        {
+            _playerService = playerService;
+        }
+
+
+        [HttpGet("GetTopTenPlayers")]
```
Comment (mmwoodfo) [inline Wordle.Api/Wordle.Api/Controllers/PlayerController.cs:21]: Typically, we don't add the action to the endpoint name.
```suggestion
        [HttpGet("TopTenPlayers")]
```

File: Wordle.Api/Wordle.Api/Controllers/PlayerController.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        public async Task<IEnumerable<Player>> GetTopTenPlayers()
+        {
+            return await _playerService.GetTopTen();
+        }
+
+        [HttpPost("AddPlayerFromBody")]
```
Comment (mmwoodfo) [inline Wordle.Api/Wordle.Api/Controllers/PlayerController.cs:27]:
```suggestion
        [HttpPost("UpsertPlayer")]
```
You may just be updating a player here too, so 'upsert' or 'update' might be a better endpoint name. Or just 'player'

Comment (VeraKorchemnaya) [reply Wordle.Api/Wordle.Api/Controllers/PlayerController.cs:27]: Thanks for your feedback Meg!

**Pull request closed without merging**